evap/staff/importers.py

```diff
@@ -1,691 +1,1281 @@
+        return {category: list(messages) for category, messages in grouped_messages}
+
+    def add_message(self, message: ImporterMessage):
+        self.messages.append(message)
+
+    def has_errors(self):
+        return any(msg.level == ImporterMessage.Level.ERROR for msg in self.messages)
+
+    def raise_if_has_errors(self):
+        if self.has_errors():
+            raise ImporterException()
+
+    def success_messages(self):
+        return self._messages_with_level_sorted_by_category(ImporterMessage.Level.SUCCESS)
+
+    def warnings_by_category(self):
+        return self._messages_with_level_by_category(ImporterMessage.Level.WARNING)
+
+    def errors_by_category(self):
+        return self._messages_with_level_by_category(ImporterMessage.Level.ERROR)
+
+    def forward_messages_to_django(self, request):
+        method_by_level = {
+            ImporterMessage.Level.SUCCESS: messages.success,
+            ImporterMessage.Level.WARNING: messages.warning,
+            ImporterMessage.Level.ERROR: messages.error,
+        }
+
+        for message in self.messages:
+            method_by_level[message.level](request, message.message)
+
+    def add_error(self, message_text, *, category=ImporterMessage.Category.GENERAL):
+        return self.add_message(ImporterMessage(ImporterMessage.Level.ERROR, category, message_text))
+
+    def add_warning(self, message_text, *, category=ImporterMessage.Category.GENERAL):
+        return self.add_message(ImporterMessage(ImporterMessage.Level.WARNING, category, message_text))
+
+    def add_success(self, message_text, *, category=ImporterMessage.Category.GENERAL):
+        return self.add_message(ImporterMessage(ImporterMessage.Level.SUCCESS, category, message_text))
+
+
+class ImporterException(Exception):
+    """Used to abort the import run immediately"""
+
+    def __init__(self, *args, message=None, category=ImporterMessage.Category.GENERAL, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.message: str = message
+        self.category: ImporterMessage.Category = category
+
+    def as_importer_message(self):
+        return ImporterMessage(ImporterMessage.Level.ERROR, self.category, self.message)
+
+
+@dataclass(eq=True, frozen=True)
+class UserData:
+    """Holds information about a user from the import file (may have to be created)"""
+
+    first_name: str
+    last_name: str
+    title: str
+    email: str
+
+    @staticmethod
+    def bulk_update_fields():
+        """Fields passed to bulk_update when updating existing users with new UserData"""
+        return ["first_name", "last_name", "title", "email", "is_active"]
+
+    def __init__(self, first_name: str, last_name: str, title: str, email: str):
+        object.__setattr__(self, "first_name", first_name.strip())
+        object.__setattr__(self, "last_name", last_name.strip())
+        object.__setattr__(self, "title", title.strip())
+        object.__setattr__(self, "email", clean_email(email))
```

Comment:
The intention was to make this usable as a key in a dictionary. For this, it needs `__eq__()` and `__hash__()`.
Maybe I should remove `eq=True` because that is true by default.

What comment would you like? Frozen tries its best to make the instances immutable by hooking setattr (I think). Immutable instances are necessary to have a safe `__hash__()` implementation, which `@dataclass` will then provide.

Reply:
I think a short "`object.__setattr__` is needed to initialize instances of frozen dataclasses" and / or a link to https://stackoverflow.com/a/58336722/13679671 or a similar reference should suffice. It's just to immediately answer the question that first readers will ask themselves; I know that I "wasted" time on finding out what this did and why because it totally caught me off guard :laughing: